pkg/controllers/disruption/consolidation.go

```diff
@@ -192,6 +193,49 @@ func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...
 		results.NewNodeClaims[0].Requirements.Add(scheduling.NewRequirement(v1beta1.CapacityTypeLabelKey, v1.NodeSelectorOpIn, v1beta1.CapacityTypeSpot))
 	}
 
+	results.NewNodeClaims[0].InstanceTypeOptions = instanceTypesOptionsFilteredByPrice
+	return Command{
+		candidates:   candidates,
+		replacements: results.NewNodeClaims,
+	}, nil
+}
+
+// Allow spot consolidation if SpotConsolidation featureflag is set to true by customers:
+// Fetch all the possible InstanceTypes for a candidate from simulation, sort it by price.
+// Only allow spot consolidation if we have atleast 15 cheapest instance types to replace the current candidate.
+// For single node consolidation, check if the current candidate is part of the first 15 cheapest instance types inorder to avoid repeated consolidation. If yes, do not consolidate.
+// If no, consolidate by sending the filtered list of instanceTypes (instanceTypesOptionsFilteredByPrice) for launch because we want to replace the current candidate with the cheaper instanceType.
+func (c *consolidation) checkForSpotToSpotConsolidation(ctx context.Context, candidates []*Candidate,
+	results *pscheduling.Results, instanceTypesOptionsFilteredByPrice []*cloudprovider.InstanceType) (Command, error) {
+
+	if !options.FromContext(ctx).FeatureGates.SpotConsolidation {
+		if len(candidates) == 1 {
+			// Spot consolidation is turned off.
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace a spot node with a spot node")...)
+		}
+		return Command{}, nil
+	}
+
+	// All possible replacements for the current candidate sorted by price
+	instanceTypesOrderedByPrice := results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions.OrderByPrice(results.NewNodeClaims[0].Requirements)
+
+	// Allow single-node/multi-node consolidation only if there are minimum number of possible InstanceTypes for a NodeClaim
+	if len(instanceTypesOrderedByPrice) < MinInstanceTypesForSpotConsolidation {
+		// InstanceType flexibility is not good enough to replace spot node.
+		c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "MinimumFlexibility of instanceTypes is not met to replace spot with spot")...)
+		return Command{}, nil
+	}
+	// For single-node consolidation, also check if the current candidate is among the 15 cheapest instance types to avoid consolidation
+	if len(candidates) == 1 {
+		_, index, found := lo.FindIndexOf(instanceTypesOrderedByPrice, func(it *cloudprovider.InstanceType) bool {
+			return it.Name == candidates[0].instanceType.Name
+		})
+		if found && index < MinInstanceTypesForSpotConsolidation {
+			return Command{}, nil
+		}
+	}
+
+	results.NewNodeClaims[0].InstanceTypeOptions = instanceTypesOptionsFilteredByPrice
```

Comment:
```suggestion
	results.NewNodeClaims[0].InstanceTypeOptions = instanceTypesOptionsFilteredByPrice[:MinInstanceTypesForSpotConsolidation]
```

We need to take a slice of our instance type options here since if we have more options than our fixed window here, we might get unexpected behavior (also, I think you only need to do this for single-node consolidation since we can keep the full array of options in multi-node since we don't care about the window size there)

Reply:
I intentionally did not slice the InstanceTypeOptions because we did not want to impact the launch path. Wouldn't slicing the array here eventually reduce it to consider only 15 instanceTypes for launch too which we wanted to avoid ?

Reply:
Discussed offline. Agreed that we need this to be sliced here, specifically for consolidation so we don't get into a continual consolidation situation where we keep thinking we can launch a cheaper instance because the Create call chooses the 16th instance type each time.

Reply:
Addressed this. Thanks.